cni/npm/ipsm/ipsm.go

```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
+	_, exists := ipsMgr.setMap[setName]
+	if exists {
+		return nil
+	}
+
+	ipsMgr.entryMap[setName] = &ipsEntry{
+		operationFlag: "-N",
+		set:           hashedName,
+		spec:          "nethash",
+	}
+	if err := ipsMgr.Run(ipsMgr.entryMap[setName]); err != nil {
+		fmt.Printf("Error creating ipset.\n")
+		return err
+	}
+
+	ipsMgr.setMap[setName] = []string{}
+
+	return nil
+}
+
+// AddToSet inserts an ip to an entry in setMap, and creates/updates the corresponding ipset.
+func (ipsMgr *IpsetManager) AddToSet(setName string, ip string) error {
+	if ipsMgr.Exists(setName, ip, "nethash") {
+		return nil
+	}
 
-	if err := ipsMgr.Create(setName); err != nil {
+	if err := ipsMgr.CreateSet(setName); err != nil {
 		return err
 	}
 
-	ipsMgr.entryMap[setName].operationFlag = "-A"
-	ipsMgr.entryMap[setName].spec = ip //This only holds one ip for now. Actually there will be multiple IPs under one setName.
+	ipsMgr.entryMap[setName].operationFlag = ipsetAppendFlag
+	ipsMgr.entryMap[setName].spec = ip
 
 	if err := ipsMgr.Run(ipsMgr.entryMap[setName]); err != nil {
 		fmt.Printf("Error creating ipset rules.\n")
 		fmt.Printf("rule: %+v\n", ipsMgr.entryMap[setName])
 		return err
 	}
+	ipsMgr.setMap[setName] = append(ipsMgr.setMap[setName], ip)
 
 	return nil
 }
 
-// DeleteFromSet removes an ip from an entry in labelMap, and delete/update the corresponding ipset.
-func (ipsMgr *IpsetManager) DeleteFromSet(setName string, ip string) (bool, error) {
-	isEmpty := false
-
-	_, exists := ipsMgr.labelMap[setName]
+// DeleteFromSet removes an ip from an entry in setMap, and delete/update the corresponding ipset.
+func (ipsMgr *IpsetManager) DeleteFromSet(setName string, ip string) error {
+	_, exists := ipsMgr.setMap[setName]
 	if !exists {
-		return false, fmt.Errorf("ipset with name %s not found", setName)
+		return fmt.Errorf("ipset with name %s not found", setName)
 	}
 
-	for i, val := range ipsMgr.labelMap[setName] {
+	for i, val := range ipsMgr.setMap[setName] {
 		if val == ip {
-			ipsMgr.labelMap[setName] = append(ipsMgr.labelMap[setName][:i], ipsMgr.labelMap[setName][i+1:]...)
+			ipsMgr.setMap[setName] = append(ipsMgr.setMap[setName][:i], ipsMgr.setMap[setName][i+1:]...)
 		}
 	}
 
-	if len(ipsMgr.labelMap[setName]) == 0 {
-		isEmpty = true
-	}
-
+	hashedName := azureNpmPrefix + util.Hash(setName)
 	entry := &ipsEntry{
-		operationFlag: "-D",
-		set:           setName,
+		operationFlag: ipsetDeletionFlag,
+		set:           hashedName,
 		spec:          ip,
 	}
-
 	if err := ipsMgr.Run(entry); err != nil {
-		fmt.Printf("Error deleing ipset entry.\n")
-		return isEmpty, err
+		fmt.Printf("Error deleting ipset entry.\n")
+		fmt.Printf("%+v\n", entry)
+		return err
 	}
 
-	return isEmpty, nil
+	return nil
 }
 
 // DeleteSet removes a set from ipset.
 func (ipsMgr *IpsetManager) DeleteSet(setName string) error {
-	entry := &ipsEntry{
-		operationFlag: "-D",
-		set:           setName,
+	_, exists := ipsMgr.setMap[setName]
```

Comment:
change to...

if _, exists := ipsMgr.setMap[setName]; !exists {
  return fmt.Errorf("ipset with name %s not found", setName)
}